### PR TITLE
added skeletonize in morphology.py

### LINKED ISSUE #2575: [Feature Request] Skeletonization
## 🚀 Feature

I would like to extract the skeleton from an image (in morphological terms, it involves reducing foreground regions in a binary image while preserving the extent and connectivity of the original region and throwing away most of the original foreground pixels), and use this in a deep learning setting.

## Motivation

Currently, there are some deep neural network-based solutions, such as DeepSkeleton, for extracting skeletons from images. While effective, they are not applicable to a wide variety of images and take up unnecessary storage and compute when there's a simpler image processing based method to solve this problem. 

## Pitch

I would like something like the ```skeletionize``` module in [scikit-image](https://scikit-image.org/docs/stable/auto_examples/edges/plot_skeleton.html) which offers two methods to perform this operation efficiently. I have attached two academic papers as references for the methods implemented in scikit-image:
1. https://www.sci.utah.edu/devbuilds/biomesh3d/FEMesher/references/lee94-3dskeleton.pdf
2. https://dl.acm.org/doi/pdf/10.1145/357994.358023

## Alternatives and Additional Context

As an alternative approach, I considered using a Python [implementation](https://gist.github.com/neeru1207/dc30df52237d5c58ded47c43ed3dcf89) by replacing the opencv functions with Kornia's morphological operations, such as erosion and opening, to solve this problem. However, I believe that having a dedicated feature would provide a faster and more user-friendly solution, akin to the efficient implementation available in scikit-image. Any help is appreciated, Thanks!

### PR DESCRIPTION
**Pull request by this-is-ac**
#### Changes

Fixes #2575

Added a vanilla implementation of skeletonization in morphology.py as described in Issue[#2575](https://github.com/kornia/kornia/issues/2575). No additional dependencies required, uses the functions `opening` and `erosion` from _morphology.py_ along with `torch`. 

#### Type of change
- [x] 🔬 New feature (non-breaking change which adds functionality)
- [x] 📝 This change requires a documentation update

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

File: kornia/morphology/morphology.py [hunk 1]
```diff
@@ -563,3 +563,55 @@ def bottom_hat(
         )
         - tensor
     )
+
+
+def skeletonize(tensor: torch.Tensor, kernel: Optional[torch.Tensor] = None) -> torch.Tensor:
+    r"""Return the skeleton of a binary image. In morphological terms, it involves reducing foreground regions in a
```
Comment (edgarriba) [inline kornia/morphology/morphology.py:569]:
```suggestion
    r"""Return the skeleton of a binary image.
    
    In morphological terms, it involves reducing foreground regions in a
```

File: kornia/morphology/morphology.py [hunk 2]
```diff
@@ -563,3 +563,55 @@ def bottom_hat(
         )
         - tensor
     )
+
+
+def skeletonize(tensor: torch.Tensor, kernel: Optional[torch.Tensor] = None) -> torch.Tensor:
```
Comment (edgarriba) [inline kornia/morphology/morphology.py:568]:
```suggestion
def skeletonize(tensor: Tensor, kernel: Tensor | None = None) -> Tensor:
```
`from __future__ import annotations`
`from kornia.core import Tensor`

File: kornia/morphology/morphology.py [hunk 3]
```diff
@@ -563,3 +563,55 @@ def bottom_hat(
...
+    r"""Return the skeleton of a binary image. In morphological terms, it involves reducing foreground regions in a
+    binary image while preserving the extent and connectivity of the original region and throwing away most of the
+    foreground pixels.
+
+    The kernel must have 2 dimensions.
+
+    Args:
+        tensor: Image with shape :math:`(B, 1, H, W)`. [C = 1]
```
Comment (edgarriba) [inline kornia/morphology/morphology.py:576]: can we add some logic to accept (B,H,W) and (H,W) /cc @johnnv1

File: kornia/morphology/morphology.py [hunk 4]
```diff
@@ -563,3 +563,55 @@ def bottom_hat(
...
+    Returns:
+        Skeleton of the image with shape :math:`(B, 1, H, W)`. [C = 1]
+
+    Example:
+        >>> tensor = torch.rand(1, 1, 20, 20)
+        >>> kernel = torch.ones(3, 3)
+        >>> skeleton_img = skeletonize(tensor, kernel)
+    """
+
+    if not isinstance(tensor, torch.Tensor):
```
Comment (edgarriba) [inline kornia/morphology/morphology.py:589]: use `KORNIA_CHECK` apis

File: kornia/morphology/morphology.py [hunk 5]
```diff
@@ -563,3 +563,55 @@ def bottom_hat(
...
+        raise TypeError(f"Kernel type is not a torch.Tensor. Got {type(kernel)}")
+
+    if len(kernel.shape) != 2:
+        raise ValueError(f"Kernel size must have 2 dimensions. Got {kernel.dim()}")
+
+    tensor = tensor.int()
+    output = torch.zeros(tensor.shape).int()
+
+    if kernel is None:
+        kernel = torch.Tensor([[0, 1, 0], [1, 1, 1], [0, 1, 0]]).int()
```
Comment (edgarriba) [inline kornia/morphology/morphology.py:605]:
```suggestion
        kernel = torch.tensor([[0, 1, 0], [1, 1, 1], [0, 1, 0]], device=tensor.device, dtype=torch.int64)
```

File: kornia/morphology/morphology.py [hunk 6]
```diff
@@ -563,3 +563,55 @@ def bottom_hat(
...
+        raise ValueError(f"Input size must have 4 dimensions. Got {tensor.dim()}")
+
+    if not isinstance(kernel, torch.Tensor):
+        raise TypeError(f"Kernel type is not a torch.Tensor. Got {type(kernel)}")
+
+    if len(kernel.shape) != 2:
+        raise ValueError(f"Kernel size must have 2 dimensions. Got {kernel.dim()}")
+
+    tensor = tensor.int()
+    output = torch.zeros(tensor.shape).int()
```
Comment (edgarriba) [inline kornia/morphology/morphology.py:602]: propagate device too

File: kornia/morphology/morphology.py [hunk 7]
```diff
@@ -563,3 +563,55 @@ def bottom_hat(
...
+        kernel = torch.Tensor([[0, 1, 0], [1, 1, 1], [0, 1, 0]]).int()
+
+    while True:
+        opened_img = opening(tensor, kernel)
+        subtracted_img = torch.sub(tensor, opened_img)
+        subtracted_img[subtracted_img < 0] = 0
+        eroded_img = erosion(tensor, kernel)
+        output = torch.bitwise_or(output, subtracted_img)
+        tensor = eroded_img.clone()
+        if torch.count_nonzero(tensor) == 0:
```
Comment (edgarriba) [inline kornia/morphology/morphology.py:614]: is this condition going to accomplish always ? or do we need to add additional?

**Review by edgarriba — COMMENTED** [review]
this needs tests too

File: kornia/morphology/morphology.py (hunk 2 above)
Comment (johnnv1) [inline kornia/morphology/morphology.py:568]: let it not be a word used so much by torch
```suggestion
def skeletonize(data: torch.Tensor, kernel: Optional[torch.Tensor] = None) -> torch.Tensor:
```

File: kornia/morphology/morphology.py (hunk 7 above)
Comment (johnnv1) [reply kornia/morphology/morphology.py:614]: shouldn't this be a for loop controlled by an iterations parameter?

File: kornia/morphology/morphology.py (hunk 3 above)
Comment (johnnv1) [reply kornia/morphology/morphology.py:576]: makes sense to me -- but this shouldn't work for C>1? since the opening and erode works

File: kornia/morphology/morphology.py (hunk 7 above)
Comment (edgarriba) [reply kornia/morphology/morphology.py:614]: Actually maybe passing bath max_iterations and max_count would be great. In case the user want to force and stop before ?

File: kornia/morphology/morphology.py (hunk 3 above)
Comment (johnnv1) [reply kornia/morphology/morphology.py]: it makes sense that it wouldn't work. When I commented yesterday, I didn't think of C as being a dimension of channels, but as N masks for a given case (batch)

**Review by johnnv1 — CHANGES_REQUESTED** [review]
@this-is-ac can you add tests and add it to the docs/?

File: test/morphology/test_skeletonize.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+            None, None, :, :
+        ]
+        assert_close(skeletonize(tensor, kernel), expected, atol=1e-4, rtol=1e-4)
+
+    def test_exception(self, device, dtype):
+        tensor = torch.ones(1, 1, 3, 4, device=device, dtype=dtype)
+        kernel = torch.ones(3, 3, device=device, dtype=dtype)
+
+        with pytest.raises(TypeError):
+            assert skeletonize([0.0], kernel)
```
Comment (johnnv1) [inline test/morphology/test_skeletonize.py:35]: In these cases, `assert` on `pytest.raises` context isn't necessary. You can match the error message by doing 
```python
       with pytest.raises(TypeError) as errinfo:
           skeletonize([0.0], kernel)
       assert 'THE ERROR MESSAGE' in str(errinfo)
```

**Review by johnnv1 — APPROVED** [review]
Thanks for the patch. I will squash your commits, cherry-pick it, and put the commit on top of our new main

File: kornia/morphology/morphology.py (hunk 7 above)
Comment (johnnv1) [reply kornia/morphology/morphology.py]: @edgarriba can you address it?

File: kornia/morphology/morphology.py [hunk 9]
```diff
@@ -563,3 +566,52 @@ def bottom_hat(
...
+
+
+def skeletonize(data: Tensor, kernel: Tensor | None = None) -> Tensor:
+    r"""Return the skeleton of a binary image.
+
+    In morphological terms, it involves reducing foreground regions in a
+    binary image while preserving the extent and connectivity of the original
+    region and throwing away most of the foreground pixels.
+
+    The kernel must have 2 dimensions.
```
Comment (edgarriba) [inline kornia/morphology/morphology.py:578]: request for subsequent pr is to add an image here as in other methods to showcase

File: kornia/morphology/morphology.py [hunk 10]
```diff
@@ -563,3 +566,52 @@ def bottom_hat(
...
+    KORNIA_CHECK(len(kernel.shape) == 2, "Kernel size must have 2 dimensions.")
+
+    while True:
+        opened_img = opening(data, kernel)
+        subtracted_img = torch.sub(data, opened_img)
+        subtracted_img[subtracted_img < 0] = 0
+        eroded_img = erosion(data, kernel)
+        output = torch.bitwise_or(output, subtracted_img)
+        data = eroded_img.clone()
+        if torch.count_nonzero(data) == 0:
```
Comment (edgarriba) [inline kornia/morphology/morphology.py:614]: how this behaves in case at least one sample of the batch hit this condition? this method looks like a candidate to work only for single images (HxW) and extend the batched version with for loop or `vmap` @johnnv1 ?

File: test/morphology/test_skeletonize.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
+import pytest
+import torch
+from torch.autograd import gradcheck
+
+from kornia.morphology import skeletonize
+from kornia.testing import assert_close
+
+class TestSkeletonize:
```
Comment (edgarriba) [inline test/morphology/test_skeletonize.py:8]:
```suggestion
class TestSkeletonize(BaseTester):
```
ideally use the common api for testing

**Review by edgarriba — APPROVED** [review]
LGTM to merge as it is and address my comments if applicable in another pull request

File: kornia/morphology/morphology.py (hunk 10 above)
Comment (johnnv1) [reply kornia/morphology/morphology.py:614]: this is true, this shouldn't work for a batched case

Comment (johnnv1) [reply kornia/morphology/morphology.py:614]: i'll merge this one, just update the docs about it, and then we open as issues the remaining items

**Review by johnnv1 — CHANGES_REQUESTED** [review]
Hey @this-is-ac i have pushed your patch [here](https://github.com/kornia/kornia/compare/main...johnnv1:kornia:feat/morph-skeletonize?expand=1). i have tested it locally and it didn't work (the stop condition never occurs) the tests even after some tests fix ups. 

If you are interested in working on it yet, you can create a new branch on your fork based on my branch, and finish it. https://github.com/johnnv1/kornia/tree/feat/morph-skeletonize

Also, needs to revert the Union typing on all morphology ops, otherwise it will not be hittable within the new typing style.

**Comment by johnnv1** [comment]
i'll be closing this PR because we are moving to the main branch, but please, if you are interested in continuing to work on this patch, you can recreate the PR from a new branch based on this branch: https://github.com/johnnv1/kornia/tree/feat/morph-skeletonize
